includes/class-wc-payments-checkout.php

```diff
@@ -126,6 +126,8 @@ public function get_payment_fields_js_config() {
 			'platformCheckoutSignatureNonce' => wp_create_nonce( 'platform_checkout_signature_nonce' ),
 			'platformCheckoutMerchantId'     => Jetpack_Options::get_option( 'id' ),
 		];
+
+		return apply_filters( 'wcpay_payment_fields_js_config', $js_config );
```

Comment:
@jrodger Do you see any issue with adding a filter to this return value? My experience with WC Pay doesn't go back too far so I want to make sure that adding a filter here doesn't have any unintended consequences.

Reply:
The only real consequence I can think of is exactly what you added it for, other code can alter the array. That's not necessarily a big problem, but it does make it easy for other plugins to alter the JS config as well. Any malicious plugin will find a way to do nasty things, but we don't necessarily want to make it easy for them.

Looking at where `wcpay_payment_fields_js_config` is used, is there any reason you need to use a filter here as opposed to adding the config line directly in this method?

cc @marcinbot, I think we've discussed similar things around filters in the past?

Reply:
> Looking at where wcpay_payment_fields_js_config is used, is there any reason you need to use a filter here as opposed to adding the config line directly in this method?

I should have been more clear in _why_ I chose to do it this way instead of adding it directly. The issue we ran into is that we need our express payment button to be able to be displayed on the product, cart, cart block, checkout, and block checkout pages. `wcpay_config` isn't normally available anywhere except the checkout page so I thought that if I'm going to be editing the content of `wcpay_config`, I need to do it everywhere it's printed to the screen. I'll say that we are printing the modified version of `wcpay_config` (which we need because a dependency of the button uses `getConfig()`) on every page that we need it, but due to script dependencies/order of output on the screen, we're unable to print our version of `wcpay_config` _after_ the default version so the default version always overwrites our custom one.

Let me know if that makes sense.

Reply:
> I feel like adding filters that are then hooked into inside the same codebase is a bit of an "antipattern".

Agreed, but it's not entirely unheard of as there are a few examples inside of WooCommerce Payments already.

> Is it possible for the `WC_Payments_Checkout` to be aware of the `WC_Payments_Platform_Checkout_Button_Handler` and then conditionally attach the `platformCheckoutButton` param here?

Possibly. Some changes would have to be made to the `WC_Payments_Platform_Checkout_Button_Handler` class, though. As of right now, the class is loaded but the `init` method returns early if the feature is not enabled. So the class always exists but we'd need a way for external code to see if it's going to do anything on that request.

> Whereas here everything is in one place, owned by us, and there should be a way to do things in a way that's easier to understand and debug in the future.

Would adding a comment above the `apply_filters()` line stating where it's being used within the codebase help? Sort of how core Woo documents where it's using its own filters/actions?

> Alternatively, it could be argued that the platform checkout is kind of an extension

These are my feelings as well. It felt "cleaner" to me to have the `WC_Payments_Platform_Checkout_Button_Handler` class contain its own logic where possible and not sprinkle it in where needed in the rest of the codebase.

Reply:
One question to this whole effort, that I haven't been able to nail down and was curious other's opinions on, is whether it would be preferable to keep all of our JS params under one umbrella (like wcpay_config) or is it preferable to break them down in to their own components as needed (like wcpay_config, wcpayPaymentRequestParams, and wcpayWooPayExpressParams)? The existing arrangement seemed to make sense for the time when there were only two (wcpay_config and wcpayPaymentRequestParams), but after I added a third I started to question it. Is there a preferred pattern to this?